presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushAggregatesIntoTableScanRuleSet.java

```diff
@@ -0,0 +1,190 @@
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import io.prestosql.Session;
+import io.prestosql.matching.Capture;
+import io.prestosql.matching.Captures;
+import io.prestosql.matching.Pattern;
+import io.prestosql.metadata.Metadata;
+import io.prestosql.metadata.Signature;
+import io.prestosql.metadata.TableHandle;
+import io.prestosql.spi.connector.AggregatePushdownResult;
+import io.prestosql.spi.connector.ColumnHandle;
+import io.prestosql.spi.connector.ProjectionApplicationResult;
+import io.prestosql.spi.expression.ConnectorExpression;
+import io.prestosql.spi.expression.Variable;
+import io.prestosql.spi.function.AggregateFunction;
+import io.prestosql.spi.type.Type;
+import io.prestosql.spi.type.TypeManager;
+import io.prestosql.sql.planner.Symbol;
+import io.prestosql.sql.planner.SymbolsExtractor;
+import io.prestosql.sql.planner.iterative.Rule;
+import io.prestosql.sql.planner.plan.AggregationNode;
+import io.prestosql.sql.planner.plan.AggregationNode.GroupingSetDescriptor;
+import io.prestosql.sql.planner.plan.TableScanNode;
+import io.prestosql.sql.tree.Expression;
+import io.prestosql.sql.tree.SymbolReference;
+import io.prestosql.type.InternalTypeManager;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+import static io.prestosql.SystemSessionProperties.isEnableAggPushdown;
+import static io.prestosql.matching.Capture.newCapture;
+import static io.prestosql.sql.planner.plan.Patterns.aggregation;
+import static io.prestosql.sql.planner.plan.Patterns.source;
+import static io.prestosql.sql.planner.plan.Patterns.tableScan;
+import static java.util.stream.Collectors.toList;
+
+public class PushAggregatesIntoTableScanRuleSet
+{
+    private static final Capture<TableScanNode> TABLE_SCAN = newCapture();
+
+    private static final Pattern<AggregationNode> AGG_WITHOUT_PROJECTION = aggregation().with(source().matching(tableScan().capturedAs(TABLE_SCAN)));
+
+    private TypeManager typeManager;
+    private final Metadata metadata;
+
+    public PushAggregatesIntoTableScanRuleSet(Metadata metadata)
+    {
+        this.metadata = metadata;
+        this.typeManager = new InternalTypeManager(metadata);
+    }
+
+    public Set<Rule<AggregationNode>> getRules()
+    {
+        return ImmutableSet.of(new PushAggregationToTableScanWithNoProjection());
+    }
+
+    private class PushAggregationToTableScanWithNoProjection
+            extends BasePushAggregationToTableScan
+    {
+        @Override
+        public Pattern<AggregationNode> getPattern()
+        {
+            return AGG_WITHOUT_PROJECTION;
+        }
+    }
+
+    private abstract class BasePushAggregationToTableScan
```

Comment:
This class hierarchy is unnecessary. There's only one implementation. Also, in general, it's better to use composition than inheritance to reuse behavior.

Reply:
I was working on the GroupingSet rule as well and the only different should be the Capture Pattern and how we extract group by columns.